Ghidra/Processors/Xtensa/data/languages/xtensaInstructions.sinc

```diff
@@ -0,0 +1,1283 @@
+}
+
+# JX - Uncoditional Jump Register, pg. 368.
+:jx as is op2 = 0 & op1 = 0 & ar = 0 & as & u2_6.7 = 0b10 & u2_4.5 = 0b10 & op0 = 0 {
+    goto [as];
+}
+
+# L8UI - Load 8-bit Unsigned, pg. 369.
+:l8ui at, as, u8_16.23 is u8_16.23 & ar = 0 & as & at & op0 = 0b0010 {
+    local addr:4 = as + zext(u8_16.23:1);
+    at = zext(*:1 addr);
+}
+
+# L16SI - Load 16-bit Signed, pg. 370.
+:l16si at, as, u9_16.23_sb1 is u9_16.23_sb1 & ar = 0b1001 & as & at & op0 = 0b0010 {
+    local addr:4 = as + zext(u9_16.23_sb1);
+    at = sext(*:2 addr);
+}
+
+# L16UI - Load 16-bit Unsigned, pg. 372.
+:l16ui at, as, u9_16.23_sb1 is u9_16.23_sb1 & ar = 0b001 & as & at & op0 = 0b0010 {
+    local addr:4 = as + zext(u9_16.23_sb1);
+    at = zext(*:2 addr);
+}
+
+# L32AI - Load 32-bit Acquire, pg. 374.
+:l32ai at, as, u10_16.23_sb2 is u10_16.23_sb2 & ar = 0b1011 & as & at & op0 = 0b0010 {
+    local addr:4 = as + zext(u10_16.23_sb2);
+    at = *:4 addr;
+    acquire(addr);
+}
+
+# L32I - Load 32-bit, pg. 378.
+:l32i at, as, u10_16.23_sb2 is u10_16.23_sb2 & ar = 0b0010 & as & at & op0 = 0b0010 {
+    local addr:4 = as + zext(u10_16.23_sb2);
+    at = *:4 addr;
+}
+
+# L32I.N - Narrow Load 32-bit, pg. 380.
+:l32i.n n_at, n_as, n_u6_12.15_sb2 is n_u6_12.15_sb2 & n_as & n_at & n_op0 = 0b1000 {
+    local addr:4 = n_as + zext(n_u6_12.15_sb2);
+    n_at = *:4 addr;
+}
+
+# L32R - Load 32-bit PC-relative, pg. 382.
+:l32r at, srel_8.23_oex_sb2 is srel_8.23_oex_sb2 & at & op0 = 0b0001 {
+    at = srel_8.23_oex_sb2;
+}
+
+# LDCT - Load Data Cache Tag, pg. 384.
+:ldct at, as is op2 = 0b1111 & op1 = 0b0001 & ar = 0b1000 & as & at & op0 = 0 {
+    at = ldct(as);
+}
+
+# LICT - Load Instruction Cache Tag, pg. 388.
+:lict at, as is op2 = 0b1111 & op1 = 0b0001 & ar = 0 & as & at & op0 = 0 {
+    at = lict(as);
+}
+
+# LICW - Load Instruction Cache Word, pg. 390.
+:licw at, as is op2 = 0b1111 & op1 = 0b0010 & ar = 0 & as & at & op0 = 0 {
+    at = licw(as);
+}
+
+# LSI - Load Single Immediate, pg. 398.
+:lsi ft, as, u10_16.23_sb2 is u10_16.23_sb2 & ar = 0 & as & ft & op0 = 0b0011 {
+    local addr:4 = as + zext(u10_16.23_sb2);
+    ft = *:4 addr;
+}
+
+# LSIU - Load Single Immediate with Update, pg. 400.
+:lsiu ft, as, u10_16.23_sb2 is u10_16.23_sb2 & ar = 0b1000 & as & ft & op0 = 0b0011 {
+    local addr:4 = as + zext(u10_16.23_sb2);
+    ft = *:4 addr;
+    as = addr;
+}
+
+# LSX - Load Single Indexed, pg. 402.
+:lsx fr, as, at is op2 = 0 & op1 = 0b1000 & fr & as & at & op0 = 0 {
+    local addr:4 = as+at;
+    fr = *:4 addr;
+}
+
+# LSXU - Load Single Indexed with Update, pg. 404.
+:lsxu fr, as, at is op2 = 0b0001 & op1 = 0b1000 & fr & as & at & op0 = 0 {
+    local addr:4 = as+at;
+    fr = *:4 addr;
+    as = addr;
+}
+
+# MADD.S - Multiply and Add Single, pg. 406.
+:madd.s fr, fs, ft is op2 = 0b0100 & op1 = 0b1010 & fr & fs & ft & op0 = 0 {
+    fr = fr f+ (fs f* ft);
+}
+
+# MAX - Maximum Value, pg. 407.
+:max ar, as, at is op2 = 0b0101 & op1 = 0b0011 & ar & as & at & op0 = 0 {
```

Comment:
With this contrived (I couldn’t find an actual binary using `min` or `max`) example:
```c
int anyltz(int a, int b, int ok) {
    int ret = 0;
    asm("min a4, %1, %2\n\t"
        "movltz %0, %3, a4\n\t"
        : "=r"(ret)
        : "r"(a), "r"(b), "r"(ok)
        : "a4");
    return ret;
}
```

I’m getting pretty much 1:1 the PCode.

My initial version:
```c
int anyltz(int param_1,int param_2,int param_3) {
  int iVar1;
  
  iVar1 = param_1;
  if (param_2 <= param_1) {
    iVar1 = param_2;
  }
  if (iVar1 < 0) {
    param_1 = param_3;
  }
  return param_1;
}
```

With your suggestion:
```c
int anyltz(int param_1,int param_2,int param_3) {
  if ((int)((uint)(param_1 < param_2) * param_1 + (uint)(param_1 >= param_2) * param_2) < 0) {
    param_1 = param_3;
  }
  return param_1;
}
```

I’m not sure this is actually more readable. Or is there anything else besides the disassembly (some CFA voodoo I haven’t found?) that’s affected by the PCode that could benefit from this?

Reply:
That's fair.  I've just noticed (and suggested by the ghidra dev's) to try to get rid of internal goto's if possible.  Could you paste the `min` code you ended up using just to verify (and `movltz` if you ended up changing that) since the second once doesn't even return param_2 ever

Reply:
I’ve used your suggestion, just with the <> swapped:
```
local cond = (as s< at) != 0;
ar = (zext(cond) * as) + (zext(!cond) * at);
```

Both don’t return param_2 (nor _1). anyltz returns param_3 if either param_1 or param_2 is < 0.